src/domain/wallets/withdrawal-fee-calculator.ts

```diff
@@ -1,10 +1,25 @@
 import { toSats } from "@domain/bitcoin"
 
-export const WithdrawalFeeCalculator = (): WithdrawalFeeCalculator => {
+export const WithdrawalFeeCalculator = ({
+  thresholdImbalance,
+  feeRatio,
+}: OnchainWithdrawalConfig): WithdrawalFeeCalculator => {
   const onChainWithdrawalFee = ({
     minerFee,
-    bankFee,
-  }: OnChainWithdrawalFeeArgs): Satoshis => toSats(minerFee + bankFee)
+    minBankFee,
+    imbalance,
+  }: {
+    minerFee: Satoshis
+    minBankFee: Satoshis
+    imbalance: SwapOutImbalance
+  }) => {
+    const aboveThreshold = Math.max(imbalance - thresholdImbalance, 0)
```

Comment:
Do I understand correctly that `imbalance` here refers to the 30 day rolling sum of the user's imbalance?

Reply:
yes @jireva. what do you think about this? it's a config so can be changed easily.

Reply:
I'm concerned that charging for imbalance this way is not ideal.  Consider a user that accumulates a large imbalance over a few days, then starts making smaller, more balanced, transactions.  Their first few transaction will pay for that imbalance but their onchain payment fees will remain high for 30 days.

(30 days itself seems reasonable. I was coding up some simulations to see how it (and other durations) would compare to tracking the full history.)

Reply:
I would recommend adding one more input to `WithdrawalFeeCalculator`: `feesPaid`(?), to track the rolling sum of fees paid by the user (in the same window as `imbalance`) and using that to make sure that the imbalance fee is not charged multiple times:
```
const aboveThreshold = Math.max(imbalance - thresholdImbalance, 0)
const bankFee = toSats(Math.max(minBankFee, Math.ceil(aboveThreshold * feeRatio) - Math.ceil(feesPaid)))
```

Another, more complex, possibility is to make the fee depend on the amount of the current transaction (that way, small transactions never get charged big fees). This requires two more inputs: `amount`, the amount of the current onchain payment and `maxFeeRatio`, the maximum fee ratio to apply to users that have an imbalance above the threshold.
And the resulting `bankFee` calculation would be something like:
```
const aboveThreshold = Math.max(imbalance - thresholdImbalance, 0)
const amountIfAboveThreshold = (aboveThreshold > 0) ? amount : 0
const bankFee = toSats(Math.max(minBankFee, Math.min(Math.ceil(amountIfAboveThreshold * maxFeeRatio), Math.ceil(aboveThreshold * feeRatio) - Math.ceil(feesPaid))))
```

Reply:
I don't think we want to add `feesPaid` because it will add a lot of overhead:
- we need to make another time consuming aggregate query to our database. also I'm not sure fees are currently simply queriable just for bankFee (I think it may be but I'm not sure)
- fee paid in absolute is not the best relevant metrics, because if a user would send 1,000 payments+ on chain in a month, even if this is small payment, there would be fees because of the min fees being paid.

but you're right on the amount. this should definitively be taken into consideration here. I'll update the formula